lib/presto-client/index.js

```diff
@@ -220,167 +229,212 @@ Client.prototype.statementResource = function(opts) {
-                 "columns": [ { "type": "bigint", "name": "cnt" } ],
-                 "infoUri": "http://localhost:8080/v1/query/20140120_032523_00000_32v8g",
-                 "id": "20140120_032523_00000_32v8g"
-             }
-            */
-            if (cancel_checker && cancel_checker()) {
-                client.request({ method: 'DELETE', path: next_uri }, function(error, code, data){
-                    if (error || code !== 204) {
-                        error_callback({message: "query fetch canceled, but Presto query cancel may fail", error: error, code: code});
-                    } else {
-                        error_callback({message: "query fetch canceled by operation"});
-                    }
-                });
+        current_req = client.request(next_uri, function(error, code, response){
+            // same as above, we have already timed out and shown an error, so abort out
+            if (timed_out) {
+                return;
+            } else if ([502, 503, 504].includes(code)) {
+                setTimeout(function(){
+                    fetch_next(next_uri);
+                }, Math.floor(Math.random() * 51) + 50);
+                if (retry_callback) {
+                    retry_callback();
+                }
                 return;
             }
-            client.request(next_uri, function(error, code, response){
-                if (error || response.error) {
+
+            if (error || (response && response.error)) {
+                if (first_request) {
+                    var message = "execution error" + (response && response.length > 0 ? ":" + response : "");
+                    if (response && response.error && response.error.message)
+                        message = response.error.message;
+                    error_callback({message, error: (error || response.error), code});
+                } else {
                     error_callback(error || response.error);
-                    return;
                 }
+                clear_timeout();
+                return;
+            }
 
-                if (state_callback && (last_state !== response.stats.state || enable_verbose_state_callback)) {
-                    state_callback(null, response.id, response.stats);
-                    last_state = response.stats.state;
-                }
+            if (first_request && (!response.id || !response.nextUri || !response.infoUri)) {
+                var error_message = null;
+                if (!response.id)
+                    error_message = "query id missing in response for POST /v1/statement";
+                else if (!response.nextUri)
+                    error_message = "nextUri missing in response for POST /v1/statement";
+                else if (!response.infoUri)
+                    error_message = "infoUri missing in response for POST /v1/statement";
+                error_callback({message: error_message, data: response});
+                clear_timeout();
```

Comment:
ditto.

Reply:
Move it up in 4f0a9aafad7b64b25d90bfa59f9777df5bd50bf7.